walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,178 @@
+    }
+
+    // Get unavailable times for mandatory attendees.
+    List<TimeRange> unavailableTimes = getUnavailableTimes(events, requestAttendees);
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+
+    // Get unavailable times for optional attendees.
+    List<TimeRange> unavailableTimesOptional = getUnavailableTimes(events, requestAttendeesOptional);
+    Collections.sort(unavailableTimesOptional, TimeRange.ORDER_BY_START);
+
+    // If there are no mandatory attendees, optional attendees are now mandatory and unavailableTimesOptional becomes empty
+    if (requestAttendees.isEmpty()) {
+      unavailableTimes = unavailableTimesOptional;
+      unavailableTimesOptional = new ArrayList<>();
+    }
+
+    return getAvailableTimes(request.getDuration(), unavailableTimes, unavailableTimesOptional);
+
   }
+
+  /**
+   * Get all unavailable times in which meeting request attendees are busy.
+   * 
+   * @param events to get unavailabilities from
+   * @param requestAttendees to find overlap for
+   * @return list of time ranges that are already occupied
+   */
+  public List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> requestAttendees) {
+    List<TimeRange> output = new ArrayList<>();
+
+    for (Event e : events) {
+      Set<String> eventAttendees = e.getAttendees();
+      
+      for (String eventAttendee : eventAttendees) {
+
+        // Checks to see if event has overlap with the attendees of the request
+        if (requestAttendees.contains(eventAttendee)) {
+          output.add(e.getWhen());
+
+          break; // Only one attendee is required for the event time to be unavailable. 
+        }
+      }
+    }
+    
+    return output;
+  }
+
+  /**
+   * Gets all available times for the meeting request attendees that contain the most available attendees.
+   *
+   * @param duration of meeting request
+   * @param unavailableTimes of mandatory attendees
+   * @param unavailableTimesOptional of optional attendees
+   * @return list of available time ranges sorted by start time
+   */
+  private List<TimeRange> getAvailableTimes(long duration, List<TimeRange> unavailableTimes, List<TimeRange> unavailableTimesOptional) {
+    List<TimeRange> output = new ArrayList<>();
+
+    // Finds all timeslots in which mandatory attendees can attend
+    List<TimeRange> attendeeTimes = getNonOverlappingTimes(duration, mergeOverlappingTimes(unavailableTimes));
+
+    // If there are no mandatory attendees or if there are no optional attendees, return attendeeTimes
+    if (unavailableTimesOptional.isEmpty()) {
+      Collections.sort(attendeeTimes, TimeRange.ORDER_BY_START);
+      return attendeeTimes;
+    }
+
+    // Find all timeslots in which optional attendees can attend
+    List<TimeRange> attendeeTimesOptional = getNonOverlappingTimes(duration, mergeOverlappingTimes(unavailableTimesOptional));
+
+    if (attendeeTimesOptional.isEmpty()) {
+      return attendeeTimes;
+    } else {
```

Comment:
Technically don't need the rest of this in an Else, since you're returning in the If above. Lmk if I missed something.

Reply:
Will update this!